src/poetry/utils/shell.py

```diff
@@ -107,6 +107,9 @@ def activate(self, env: VirtualEnv) -> int | None:
             if self._name == "zsh":
                 # Under ZSH the source command should be invoked in zsh's bash emulator
                 c.sendline(f"emulate bash -c '. {shlex.quote(str(activate_path))}'")
+        elif self._name == "fish":
+            # Under fish "\r" should be sent explicitly
+            c.sendline(f"source {shlex.quote(str(activate_path))}\r")
         else:
             c.sendline(
                 f"{self._get_source_command()} {shlex.quote(str(activate_path))}"
```

Comment:
```suggestion
        else:
            cmd = f"{self._get_source_command()} {shlex.quote(str(activate_path))}"
            if self._name == "fish":
                # Under fish "\r" should be sent explicitly
                cmd += "\r"
            c.sendline(
                cmd
```

and revert the change in `_get_source_command()`